pkg/client/connection.go

```diff
@@ -30,4 +30,8 @@ type Connection interface {
 	// Close closes the connection, releasing all the resources that it uses. Once closed the
 	// connection can't be reused.
 	Close() error
+
+	// Publish / Subscribe interface
+	Publish(m Message, topic string) error
+	Subscribe(topic string, callback func(m Message, topic string) error) error
```

Comment:
Same as before

Reply:
Same for the subscriber, it should be a separate `Subscriber` type.

I think that the topic does not need to be a parameter of the callback, the only parameter should be the message.

Reply:
Another reason to have these separate objects is that then they can have additional methos. For example, you may want to stop receiving messages from a topic. How do you do that? If you have a `Subscriber` type then it can have a `Close` method:

```go
subscriber, err := connection.NewSubscriberBuilder().
   ...
   Build()
if err != nil {
  ...
}
defer subscriber.Close()
```

Reply:
As a user would I have an unsubscribe? Asking in the context of what I have to do when shutting down my service (`cluster-service`)

Reply:
When shutting down the service you will not be removing the subscription, because when you restart it you will want to receive the messages that were published when you were down.

Note that doing something like `subscriber.Close()` doesn't remove the subscription, it just tells the library that you don't want to receive the messages now. They will still be saved by the message broker (if the subscription is _durable_), and delivered later, whenever to create the subscriber again.

So, in an ordered shutdown of your service you will want to call `subscriber.Close()` because you don't want to incoming messages to be processed in the middle of your shutdown process.